drivers/timer/stm32_lptim_timer.c

```diff
@@ -0,0 +1,285 @@
+		LL_LPTIM_SetAutoReload(LPTIM1, LPTIM_TIMEBASE);
+
+		/* do not use compare yet, set the compare to period */
+		LL_LPTIM_SetCompare(LPTIM1, 0);
+
+		/* Start the LPTIM counter in continuous mode */
+		LL_LPTIM_StartCounter(LPTIM1,
+				LL_LPTIM_OPERATING_MODE_CONTINUOUS);
+	} else {
+		/* Enable the LPTIM1 counter */
+		LL_LPTIM_Enable(LPTIM1);
+
+		/* Set the Autoreload value once the timer is enabled */
+		LL_LPTIM_SetAutoReload(LPTIM1, COUNT_PER_TICK);
+
+		/* do not use compare yet, set the compare to period */
+		LL_LPTIM_SetCompare(LPTIM1, 0);
+
+		/* Start the LPTIM counter in continuous mode */
+		LL_LPTIM_StartCounter(LPTIM1,
+				LL_LPTIM_OPERATING_MODE_CONTINUOUS);
+	}
+
+	/* poll ARROK bit to validate write operation to ARR register */
+	while (LL_LPTIM_IsActiveFlag_ARROK(LPTIM1) == 0) {
+	}
+	LL_LPTIM_ClearFlag_ARROK(LPTIM1);
+
+#ifdef CONFIG_DEBUG
+	/* stop LPTIM1 during DEBUG */
+	LL_DBGMCU_APB1_GRP1_FreezePeriph(LL_DBGMCU_APB1_GRP1_LPTIM1_STOP);
+#endif
+	return 0;
+}
+
+void z_clock_set_timeout(s32_t ticks, bool idle)
+{
+	/* LPTIM1 AutoReload value aligned on Kernel ticks */
+	u32_t next_arr = 0;
+
+	ARG_UNUSED(idle);
+
+#ifdef CONFIG_TICKLESS_KERNEL
+	if (ticks == K_FOREVER) {
+		/* disable LPTIM */
+		LL_APB1_GRP1_ForceReset(LL_APB1_GRP1_PERIPH_LPTIM1);
+		LL_APB1_GRP1_DisableClock(LL_APB1_GRP1_PERIPH_LPTIM1);
+		return;
+	}
+
+	/* passing ticks==1 means "announce the next tick",
+	 * ticks value of zero (or even negative) is legal and
+	 * treated identically: it simply indicates the kernel would like the
+	 * next tick announcement as soon as possible.
+	 */
+	if (ticks <= (s32_t)1) {
+		ticks = 1;
+	} else {
+		ticks = (ticks - 1);
+	}
+	/* maximise Tick to keep next_arr on 32bit values
+	 * in anycase the ARR cannot exceed LPTIM_TIMEBASE
+	 */
+	if (ticks > (s32_t)0xFFFF) {
+		ticks = 0xFFFF;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	/* read current counter value (cannot exceed 16bit) */
+	u32_t cnt_value = LL_LPTIM_GetCounter(LPTIM1) + 1;
+	/* calculate the next arr value (cannot exceed 16bit)
+	 * adjust the next ARR match value to align on Ticks
+	 * from the current counter value to first next Tick
+	 */
+	next_arr = (((cnt_value * (CONFIG_SYS_CLOCK_TICKS_PER_SEC))
+			/ (LPTIM_CLOCK)) + 1) * (LPTIM_CLOCK)
+			/ (CONFIG_SYS_CLOCK_TICKS_PER_SEC);
+	/* add ms from the expected nb of Ticks */
+	next_arr = next_arr + ((u32_t)(ticks) * LPTIM_CLOCK)
+			/ CONFIG_SYS_CLOCK_TICKS_PER_SEC;
+
+	/* minimise (ie 10 counter cycles) to trig timer */
+	if ((next_arr - LL_LPTIM_GetAutoReload(LPTIM1)) < 10) {
+		next_arr = LL_LPTIM_GetAutoReload(LPTIM1) + 10;
+	}
+
+	/* maximise to TIMEBASE */
+	if (next_arr > LPTIM_TIMEBASE) {
+		next_arr = LPTIM_TIMEBASE;
+	}
+	/* run timer and wait for the reload match */
+	LL_LPTIM_SetAutoReload(LPTIM1, next_arr);
+	/* poll ARROK bit to validate write operation to ARR register */
+	while (LL_LPTIM_IsActiveFlag_ARROK(LPTIM1) == 0) {
```

Comment:
This loop can induce a very latency, by doing so you will always loop and "waste" time here
maybe it would be better to add this check at the beginning of the function to verify that the previous programmin was okay and therefore to wait only in case the previous programming has not completed yet ..  ?